perf(transformer): optimize inserting var/let statements

Dunqing opened the pull request:
Found while working on #9881, this PR aims to optimize inserting statements by reducing an unnecessary `Vec` allocation.

Previously, `get_var_statement` would collect `var` statement and `let` statement into a `std::vec::Vec` and return it, then immediately `extend`/`insert` to another `ArenaVec`. Considering here only have two elements, we don't need to store them in a `Vec`, just return them, and use them depending on how they are used.

Seems to have very minor improvement, and I guess `transformer/checker.ts` has 0.1 regression is noisy.

<img width="767" alt="image" src="https://github.com/user-attachments/assets/c0aeeb5a-e2da-4e2a-9990-56b239f6ab65" />

Dunqing commented:
* **#10654** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10654?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10654?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10654).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (approved):
It's definitely an improvement not to allocate a temp `Vec`.

What is less good is that we always copy the whole of `stmts` into a new `ArenaVec`. If `stmts` has spare capacity, it'd be preferable to shift up the existing `Statement`s to make room for the new ones at the start, and then move the new statements into the newly-created space. That doesn't reduce the amount of memory copying that needs to happen, but at least it does it without consuming more memory in arena for a brand new `Vec`.

However, I was surprised to see that `Vec::splice` was doing the same unoptimal thing - allocate a new `Vec` and copy everything over. `Vec::insert` is also surprisingly bad! If the `Vec` is already full to capacity, it copies all the existing elements of the `Vec` *twice*, which is completely unnecessary.

To get this (and the many similar operations in our code) optimal, we'd need new APIs like:

```rs
impl<'a, T> Vec<'a, T> {
    /// Same as `vec.insert(0, value)`
    fn push_start(&mut self, value: T);
    /// Same as `vec.splice(0..0, iter)`
    fn extend_start<I: IntoIterator<Item = T>>(&mut self, iter: I);
}
```

We could optimize these to use the "shift up" method.

overlookmotel commented:
### Merge activity

* **Apr 29, 12:00 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
perf(transformer): optimize inserting var/let statements (#10654)

Found while working on #9881, this PR aims to optimize inserting statements by reducing an unnecessary `Vec` allocation.

Previously, `get_var_statement` would collect `var` statement and `let` statement into a `std::vec::Vec` and return it, then immediately `extend`/`insert` to another `ArenaVec`. Considering here only have two elements, we don't need to store them in a `Vec`, just return them, and use them depending on how they are used.

Seems to have very minor improvement, and I guess `transformer/checker.ts` has 0.1 regression is noisy.

<img width="767" alt="image" src="https://github.com/user-attachments/assets/c0aeeb5a-e2da-4e2a-9990-56b239f6ab65" />
